torch/nn/parameter.py

```diff
@@ -42,3 +46,49 @@ def __reduce_ex__(self, proto):
             torch._utils._rebuild_parameter,
             (self.data, self.requires_grad, OrderedDict())
         )
+
+
+class _UninitializedParameter(Parameter):
+    r"""A parameter that is not yet initialized for shape inference support.
+
+    UninitializedParameters are detected by the optimizer ...
+    """
+    def detach(self):
```

Comment:
This looks weird that you have to do that. Where does it come from?

Reply:
The problem here was that creating the `state_dict` ends up calling `detach` on the parameter and returning a tensor object. 

This is pretty hackish and probably a check in the `state_dict` is cleaner a better.

Reply:
changed it in `state_dict` and clean up the `_UnitializedParameter` code.